### DISCUSSION TITLE
Is it possible to obtain a unique identifier using this library, that works across devices?

### DISCUSSION THREAD

**Discussion by alorma** [vestrel00/contacts-android#194] [Q&A] [answered]
Hello!

Is it possible to obtain a unique identifier using this library, that works across devices?

My use case is to create an app to automatically delete contacts after a given time (temporal contacts), and I wanted to do it online so user can keep it on multiple devices...

Is it even possible? Is there a unique "identifier" for contact across devices?

**Comment by vestrel00** [comment]
Hey @alorma 👋 Thanks for asking this question! I need to do a bit of research first before I give you a concrete answer.

I think #39 will help answer this question. I'll work on it as soon as I can!

The closest thing to this "ID that could be used to identify contacts across devices" is https://developer.android.com/reference/android/provider/ContactsContract.ContactsColumns#LOOKUP_KEY because it has to do with syncing and aggregation. I created issue #140 to add support for it. I just included in the [v0.1.10 Release Checklist](https://github.com/vestrel00/contacts-android/discussions/137) 🔥 

What I can say right now is that anything that the Android Open Source Project (AOSP) Contacts app (the Contacts app that comes with every copy of Android) can do, this library will also allow you to do. If AOSP Contacts can do something that this library can't do for you, let me know and I will make sure that there is an issue for it so I can implement it 😁 

------------------------------------

In the meantime, @alorma could you tell me the exact steps you want to accomplish? Is it something like this...?

1. Create Contact.
2. Sync it to make it available across devices.
3. After a few minutes, hours, or days, automatically delete the contact across devices.

If this is what you want to do, then you can already do it with this library. 

But first, keep in mind that depending on the Account sync settings in the device, any Contact you create on one device will be available on all your devices that are logged in to the same Account (e.g. alorma@gmail.com) with Contact syncing enabled. The same thing applies for when you delete a Contact. Deleting a Contact on one device will also delete it in other devices. But, all of this depends on the system sync settings for the account for each device. Users can be offline or they turned off Contact syncing. A lot of things can go wrong. Contacts are not guaranteed to be synced at all times.

Another thing you should know is that Contact data is synced across all devices via sync adapters interfacing with a remote backend server. In the case of google accounts (e.g. alorma@gmail.com), Google sync adapters in the device manage syncing Contact data across devices by getting and setting data from/to Google servers. Google servers are the source of truth and contain all Google Contacts. This library does not provide a way for you to interact with Google backends servers or any other backends. 

> All of this documentation will be added in #39.

Anyways, if you are still looking to implement a feature that automatically deletes contacts created after a certain amount of time... Using this library, it might look something like...

```kotlin
fun createTemporaryContact(newRawContact: NewRawContact, deleteOnDate: Date) {
    val contactsApi = Contacts(context)
    val insertResult = contactsApi.insert().contacts(newContact).commit()
    val contactId = insertResult.contact(contactsApi, newRawContact)
    deleteContactOnDate(contactId, deleteOnDate)
}

fun deleteContactOnDate(contactId: Long, deleteOnDate: Date) {
    // Use work manager or something to schedule running a job to delete the contact on the given date.
}
```

Once #140 has been implemented, you should use the Contact `lookupKey` instead of the `id` because it is more reliable. While the Contact `id` or `lookupKey` may not be the same across devices, the Contacts Provider will handle all the syncing. So if you insert, update, or delete a contact on one device, other devices will also see that change. The `id` or `lookupKey` does not need to be the same across devices.

Notice that this is all client-side code. It's all code in your Android app/service. Syncing changes (inserts, deletes, updates) across devices are done automatically by the Contacts Provider, depending on Account Contact sync settings.

**Comment by alorma** [comment]
Cool! That is helpful!

I guessed that it worked that way on sync across device.

My concer was If i create a contact and save on some online database for my app (example: Firebase) so I can get the same data on mutiple devices.

BTW, I will start interacting only on local so for now I will not worry much on syncing, just exploring the possibilities for now!

Thanks

**Comment by vestrel00** [comment]
> save on some online database for my app (example: Firebase) so I can get the same data on multiple devices.

I would highly advice you NOT to do this. The Contacts Provider along with Account sync adapters manage all Contact data in their own online database. If you save a copy of Contacts in your own database, it will quickly get out of sync with the Account online database.

> I will start interacting only on local so for now I will not worry much on syncing

This library only documents the syncing behavior. It currently does not provide syncing functions. You would need to implement your own sync adapters, which is currently not in scope of this library.

I'll work on #39 before I close this issue. As I mentioned, you can use the `Contact.id` for now but should migrate to `Contact.lookupKey` once #140 is implemented in the next release.

**Comment by alorma** [comment]
Cool, thanks

**Comment by vestrel00** [comment, marked as answer]
@alorma, I did some testing. It seems like the **`LOOKUP_KEY` is the same across devices** as long as data has been synced (up-to-date) with the servers!

I will close this issue once I've completed;

- #140 
- #39 

Thanks for raising this issue!

**Comment by vestrel00** [comment]
I'll actually close this issue now that I've completed documentation in #39. Keep an eye out for the progress on #140 if you are interested in using lookup keys 😁

**Comment by alorma** [comment]
Thanks!
